kernel/safe_typing.ml

```diff
@@ -1332,33 +1267,115 @@ let export ?except ~output_native_objects senv dir =
   in
   mp, lib, ast
 
-(* cst are the constraints that were computed by the vi2vo step and hence are
- * not part of the mb.mod_constraints field (but morally should be) *)
-let import lib cst vodigest senv =
+(** Assumes that the global constraints of the module are already present *)
+let really_import lib vodigest senv =
   check_required senv.required lib.comp_deps;
   check_engagement senv.env lib.comp_enga;
   if DirPath.equal (ModPath.dp senv.modpath) lib.comp_name then
     CErrors.user_err ~hdr:"Safe_typing.import"
      (Pp.strbrk "Cannot load a library with the same name as the current one.");
   let mp = MPfile lib.comp_name in
   let mb = lib.comp_mod in
-  let env = Environ.push_context_set ~strict:true
-                                     (Univ.ContextSet.union mb.mod_constraints cst)
-                                     senv.env
-  in
+  let env = senv.env in
   let env =
     let linkinfo = Nativecode.link_info_of_dirpath lib.comp_name in
     Modops.add_linked_module mb linkinfo env
   in
   let env = Environ.add_native_symbols lib.comp_name lib.comp_natsymbs env in
+  let sections =
+    Option.map (Section.map_custom (fun custom ->
+        {custom with rev_reimport = (lib, vodigest) :: custom.rev_reimport}))
+      senv.sections
+  in
   mp,
   { senv with
     env;
     modresolver = Mod_subst.add_delta_resolver mb.mod_delta senv.modresolver;
     required = DPmap.add lib.comp_name vodigest senv.required;
     loads = (mp,mb)::senv.loads;
+    sections;
   }
 
+(* cst are the constraints that were computed by the vi2vo step and hence are
+ * not part of the mb.mod_constraints field (but morally should be) *)
+let import lib cst vodigest senv =
+  let senv = push_context_set ~strict:true lib.comp_mod.mod_constraints senv in
```

Comment:
Using Safe_typing.push_context_set means we add the dependency's universes to the current module's universes, that seems wrong to me.

Reply:
You're right indeed.

Reply:
That said, if I understand the code correctly, nested modules transitively contain the universe constraints of their subcomponents. Really, this global state shouldn't be stored this way.